src/net.h

```diff
@@ -666,6 +666,12 @@ class NetEventsInterface
     /** Handle removal of a peer (clear state) */
     virtual void FinalizeNode(const CNode& node) = 0;
 
+    /**
+     * Callback to determine whether the given set of service flags are sufficient
+     * for a peer to be "relevant".
+     */
+    virtual bool HasDesirableServiceFlags(ServiceFlags services) = 0;
```

Comment:
> This interface is not the right place for a callback like this because it is not a "net event".

Hmm ok, that would avoid locking `cs_main` in the open connections thread. Which is nice.

Reply:
> Because, most of the time, the peer manager class will call to the CConMan setter and subsequently call the getter to utilize it. 

I wasn't saying to make connman the container for this data, I only meant to add a setter to inform connman of the things it needs to perform its job (making connections to desirable nodes). The main container for the flags would still be peerman.

Reply:
> > Because, most of the time, the peer manager class will call to the CConMan setter and subsequently call the getter to utilize it.
> 
> I wasn't saying to make connman the container for this data, I only meant to add a setter to inform connman of the things it needs to perform its job (making connections to desirable nodes). The main container for the flags would still be peerman.

Ok. What I dislike about that approach is that we will lose a lot of flexibility.

The PeerManager should be able to decide granularly whether to connect to a specific address before ConnMann opens a connection and initiates the handshaking protocol.
The node's contextual information and logic required to take the decision of opening a socket to a particular address resides mostly in PeerManager (right now, we only have the stale chain condition, but there could be more). And, by adding another field inside ConnMan, we would be duplicating data (same `desirable_services` field inside PeerManager and ConnMan), which would need to be synchronized and could potentially lead to race conditions.

I mean, this misalignment could result in ConnMan performing actions that are no longer in line with the latest PeerManager state. For instance, it might create connections to multiple peers with certain service flags, when PeerManager only intended to connect to one peer supporting such service flags. This situation would eventually lead to PeerManager closing the newly created connections when it sends the version message, as these connections are no longer desirable. Which is a waste of resources.

Thus why I think that the callback mechanism suits well here. So PeerManager is able to evaluate each connection request, and avoid executing the connection + handshake when we are going to disconnect the peer anyway.

Reply:
Circling back to this, I still don't think the interface choices in this PR are right.

Mostly it's just that the `NetEventsInterface` is not the place for a getter like this. It is meant for networking events (hence the name) such as opening/closing connections and processing/sending messages. So Approach NACK on misusing this interface.

> The node's contextual information and logic required to take the decision of opening a socket to a particular address resides mostly in PeerManager (right now, we only have the stale chain condition, but there could be more)

This is not true? [`ThreadOpenConnections`](https://github.com/bitcoin/bitcoin/blob/03c5b0064d4f766bc8dc6508773c7579e9ad39bc/src/net.cpp#L2415) is a method on `CConnman` and handles the opening of network connections.

I would suggest that you take the same approach as we have already taken with `CConnman::{SetTryNewOutboundPeer, StartExtraBlockRelayPeers}`, i.e. peerman telling connman what to do. That approach avoids increasing the coupling between the two components.

Reply:
> > The node's contextual information and logic required to take the decision of opening a socket to a particular address resides mostly in PeerManager (right now, we only have the stale chain condition, but there could be more)
> 
> This is not true? [`ThreadOpenConnections`](https://github.com/bitcoin/bitcoin/blob/03c5b0064d4f766bc8dc6508773c7579e9ad39bc/src/net.cpp#L2415) is a method on `CConnman` and handles the opening of network connections.

Yes, that's the layering violation that causes the coupling between PeerManager and ConMann that we both dislike and shouldn't be there (at least not entirely). The connections manager alone does not have enough information to select the best connection candidate based on the node's context. CConMan cannot access the chain state, the sync progress, in-flight block requests, nor any services predilection, nor any other user customization to prefer certain peers over others—such as ones supporting a number of compact block filters or v2 encrypted connections, etc. This is the reason why we are forced to set the bool flags you mentioned (which can easily get out of sync, causing a waste of resources and races). Under this structure, the only thing we do is continue expanding the number of fields inside CConMan that are susceptible to the same issues (see below).

> I would suggest that you take the same approach as we have already taken with CConnman::{SetTryNewOutboundPeer, StartExtraBlockRelayPeers}, i.e. peerman telling connman what to do. That approach avoids increasing the coupling between the two components.

I still stand by what I wrote previously at https://github.com/bitcoin/bitcoin/pull/28170#discussion_r1297413491.
Essentially, in my view, the approach is resource-wasteful and could lead to race conditions. And, following what I wrote above, it has a poor structure that increases the number of unrelated fields inside CConMan over time.

Overall, I'm in favor of decoupling these two components, but I'm not in favor of continuing to expand the current structure. The connections' desirable services flag can (and will) change continually over time, depending on the node's state, connection heuristics, and possible user custom configs. It is a much more dynamic field than the ones you mention.

> Mostly it's just that the NetEventsInterface is not the place for a getter like this. It is meant for networking events (hence the name) such as opening/closing connections and processing/sending messages.

Two points about this:

Firstly, it is not just a getter. The node needs to make the decision to connect to a certain address based on several factors that depend on the node's context. Caching a flag doesn't help here. I mentioned some possible use cases above. I don't think continuing to expand CConMan fields with data from upper layers is a good structure.

Secondly, I believe that with this statement, we are somewhat in agreement that the logic of "to which address, supporting which services, should we connect?" shouldn't be part of CConMan. However, as long as it remains there, the process of deciding to connect (or not) to a certain address is a net event. Like you, I would prefer if it weren't there, and we could work on it. But I don't think this PR is introducing something incorrect right now.

Reply:
> CConMan cannot access the chain state, the sync progress, in-flight block requests

I don't see how the connection opening logic needs access to any of those things.

> number of compact block filters

Bitcoin Core doesn't download compact block filters.

> v2 encrypted connections, etc.

Connman is doing this right now using our service flags and the service flag of the potential new peer. It has all the data it needs.

> Essentially, in my view, the approach is resource-wasteful and could lead to race conditions.

Your approach has the exact same "flaw" w.r.t. connecting to desired service flags, e.g.:

* connman calls `HasDesireableServiceFlags` for peer A which returns true
* connman proceeds with opening the connection to peer A
* some event occurs (e.g. messages from other peers are processed) that would make `HasDesireableServiceFlags` return false for peer A
* connection to peer A is established

Peer A is now connected even-though `HasDesireableServiceFlags` would return false. These races can't really be avoided, so it's best to stick with the interface design we've already got imo.